pkg/manager/controllers/addontemplate/controller.go

```diff
@@ -0,0 +1,161 @@
+	addonInformers addoninformers.SharedInformerFactory,
+	clusterInformers clusterv1informers.SharedInformerFactory,
+	dynamicInformers dynamicinformer.DynamicSharedInformerFactory,
+) factory.Controller {
+	c := &addonTemplateController{
+		kubeConfig:       hubKubeconfig,
+		hubKubeClient:    hubKubeClient,
+		addonClient:      addonClient,
+		cmaLister:        cmaInformer.Lister(),
+		addonManagers:    make(map[string]context.CancelFunc),
+		addonInformers:   addonInformers,
+		clusterInformers: clusterInformers,
+		dynamicInformers: dynamicInformers,
+	}
+
+	return factory.New().WithInformersQueueKeysFunc(
+		func(obj runtime.Object) []string {
+			key, _ := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+			return []string{key}
+		},
+		cmaInformer.Informer()).
+		WithSync(c.sync).
+		ToController("addon-template-controller")
+}
+
+func (c *addonTemplateController) stopUnusedManagers(
+	ctx context.Context, syncCtx factory.SyncContext, addOnName string) {
+
+	stopFunc, ok := c.addonManagers[addOnName]
+	if ok {
+		stopFunc()
+	}
+}
+
+func (c *addonTemplateController) sync(ctx context.Context, syncCtx factory.SyncContext, key string) error {
+	_, addonName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		// ignore addon whose key is not in format: namespace/name
+		return nil
+	}
+
+	cma, err := c.cmaLister.Get(addonName)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+
+	if !utils.SupportAddOnTemplate(cma) {
+		c.stopUnusedManagers(ctx, syncCtx, cma.Name)
+		return nil
+	}
+
+	_, exist := c.addonManagers[addonName]
+	if exist {
+		klog.Infof("There already is a manager started for addon %s, skip.", addonName)
+		return nil
+	}
+
+	klog.Infof("Start an addon manager for addon %s", addonName)
+
+	stopFunc := c.startManager(ctx, addonName)
+	c.addonManagers[addonName] = stopFunc
+	return nil
+}
+
+func (c *addonTemplateController) startManager(
+	pctx context.Context,
+	addonName string) context.CancelFunc {
+	ctx, stopFunc := context.WithCancel(pctx)
+	go func() {
+		err := c.runController(ctx, addonName)
+		if err != nil {
+			klog.Errorf("run controller for addon %s error: %v", addonName, err)
+			utilruntime.HandleError(err)
+		}
+	}()
+	return stopFunc
+}
+
+func (c *addonTemplateController) runController(
+	ctx context.Context, addonName string) error {
+	mgr, err := addonmanager.New(c.kubeConfig)
+	if err != nil {
+		return err
+	}
+
+	agentAddon := addonfactory.NewCRDTemplateAgentAddon(
+		addonName,
+		c.hubKubeClient,
+		c.addonClient,
+		addonfactory.GetAddOnDeploymentConfigValues(
+			addonfactory.NewAddOnDeploymentConfigGetter(c.addonClient),
+			addonfactory.ToAddOnCustomizedVariableValues,
+			addonfactory.ToAddOnNodePlacementPrivateValues,
+			addonfactory.ToAddOnRegistriesPrivateValues,
+		),
+	)
+	err = mgr.AddAgent(agentAddon)
+	if err != nil {
+		return err
+	}
+
+	err = mgr.StartWithInformers(ctx, c.addonInformers, c.clusterInformers, c.dynamicInformers)
```

Comment:
how about work informer? it is the heaviest one I think

Reply:
the workinformer and kubeinformer used [here](https://github.com/open-cluster-management-io/addon-framework/blob/ebc4aa86b40ce5ae6e3104766cc3b0c42453ba5f/pkg/addonmanager/manager.go#L120-L129) are filtered by addon name. so I think it is not necessary to share.